src/main/java/org/killbill/billing/plugin/dwolla/core/DwollaNotificationHandler.java

```diff
@@ -0,0 +1,193 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.killbill.billing.plugin.dwolla.core;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import io.swagger.client.model.HalLink;
+import io.swagger.client.model.Webhook;
+import org.joda.time.DateTime;
+import org.killbill.billing.account.api.Account;
+import org.killbill.billing.account.api.AccountApiException;
+import org.killbill.billing.osgi.libs.killbill.OSGIKillbillAPI;
+import org.killbill.billing.payment.api.*;
+import org.killbill.billing.payment.plugin.api.PaymentPluginApiException;
+import org.killbill.billing.payment.plugin.api.PaymentPluginStatus;
+import org.killbill.billing.plugin.dwolla.api.DwollaCallContext;
+import org.killbill.billing.plugin.dwolla.api.DwollaPaymentPluginApi;
+import org.killbill.billing.plugin.dwolla.client.DwollaClient;
+import org.killbill.billing.plugin.dwolla.dao.DwollaDao;
+import org.killbill.billing.plugin.dwolla.dao.gen.tables.records.DwollaResponsesRecord;
+import org.killbill.billing.plugin.dwolla.util.JsonHelper;
+import org.killbill.billing.util.callcontext.CallContext;
+import org.killbill.billing.util.callcontext.TenantContext;
+import org.killbill.clock.Clock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import java.sql.SQLException;
+import java.sql.SQLIntegrityConstraintViolationException;
+import java.util.UUID;
+
+import static org.killbill.billing.plugin.dwolla.client.EventTopic.valueOf;
+
+public class DwollaNotificationHandler {
+
+    private static final Logger logger = LoggerFactory.getLogger(DwollaNotificationHandler.class);
+
+    private final DwollaDao dao;
+    private final DwollaClient client;
+    private final OSGIKillbillAPI osgiKillbillAPI;
+    private final Clock clock;
+
+    public DwollaNotificationHandler(final DwollaDao dao, final DwollaClient client, final OSGIKillbillAPI osgiKillbillAPI, final Clock clock) {
+        this.dao = dao;
+        this.client = client;
+        this.osgiKillbillAPI = osgiKillbillAPI;
+        this.clock = clock;
+    }
+
+    public void processNotification(final String notification, final UUID tenantId) throws PaymentPluginApiException {
+        final Webhook webhook = JsonHelper.getObjectFromRequest(notification, Webhook.class);
+        try {
+            dao.addNotification(webhook, clock.getUTCNow(), tenantId);
+        } catch (Exception e) {
+            if (e.getCause() instanceof SQLIntegrityConstraintViolationException) {
+                logger.debug("Dwolla notification was already processed");
```

Comment:
In that case, we usually keep duplicates in this table (no constraint) and only look at the latest entry (among other things, it lets us monitor how the provider is behaving w.r.t. retries for instance).

No need to change it though - we'd need to gather production data first anyways to see if it's a common occurrence.